File: sliding_window/python/min_subarray.py
def subArray(s, array):
    left = 0
    right = 0
    total = 0
    array_len = len(array)
    min_len = array_len + 1
    print s 
    while (right < array_len):
        total += array[right] 
        right+=1
        if(total >= s):
            while(left < right and total - array[left] >= s):
                total -= array[left]
                left+=1

        if(total >= s and min_len > (right - left)):
            min_len = right - left;    
    if(min_len <= array_len):
        print "sub array is",
        for i in range(left, right):
            print array[i],
        print
        return min_len 
    else:
        return 0

list = [2,3,1,4,3,5,7]
s = 15
minlen = subArray(s, list)
print "min len is",minlen

File: sliding_window/c/min_subarray.c
#include <stdio.h>

int min_subarray_len(int *left, int *right, int s, int *array, int len)
{
    *left = 0, *right = 0;
    // if s is greater than the sum of all value in array, min_len should be greater than len forever
    int min_len = len + 1, array_len = len, total = 0;
    while(*right < array_len)
    {
       do
       {
           total += array[*right];
           ++(*right); 
       }
       while(*right < array_len && total < s);       

       while(*left < *right && total - array[*left] >= s)
       {
           total -= array[*left];
           ++(*left);
       }

       // Only change the value of min_len while s is less than the sum of all value in array 
       if(total >=s && min_len > (*right - *left) )
       {
           min_len = *right - *left;
       }
    }

    return min_len <= len ? min_len : 0;
}

int main()
{
    int array[] = {2,3,1,4,3,5,7};
    int left = 0, right = 0, s = 22; 
    int mlen = min_subarray_len(&left, &right, s, array, 7);
    printf("min len: %d\n", mlen);
    for(int i = left; i < right; ++i)
        printf("%d, ", array[i]);
    printf("\n");
    return 0;
}
